### PR TITLE
ENH: Add _argmaxima1d for maxima detection in 1D arrays

### PR DESCRIPTION
**Pull request by lagru**
This new Cython function is intended to be wrapped by a peak finding function proposed in #8264. It finds all local maxima in a 1D array. 

### Motivation

There already exists a similar function in SciPy: [scipy.signal.argrelmax](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.argrelmax.html). However  it has the following disadvantages compared to the proposed function:

* It can't detect flat maxima (more than one sample wide).
* it is "relatively" slow compared to this Cython implementation 

Please see this [comparison](http://nbviewer.jupyter.org/urls/gitlab.com/snippets/1696398/raw) for a demonstration of these points.

### Summary of changes

* Added `_peak_finding_utils.pyx` with the new function `_argmaxima1d`
* Added appropriate unit tests in `test_peak_finding.py`.
* Added instruction to compile new extension in appropriate files.
* Added generated c-file to gitignore.

### To do & open questions

* <strike>The current version requires the input array to have the dtype `np.float64` and requires conversion if it is used with other dtypes such as (unsigned) integers. Using Cythons `ctypedef fused` statement this function could be made more flexible at the cost of an increased binary. What is the preferred behavior here?</strike>

### Related discussions & links

* Pull Request #8264.
* [Performance comparison](http://nbviewer.jupyter.org/urls/gitlab.com/snippets/1696398/raw) with `scipy.signal.argrelmax`
* Related [discussion](https://mail.python.org/pipermail/scipy-dev/2018-January/022405.html) on SciPy-Dev mailing list

--

Thank you all in advance for taking the time to review this PR and any feedback given.

### CODE REVIEW

**Comment by lagru** [comment]
Ping @larsoner

File: scipy/signal/_peak_finding_utils.pyx
Comment (larsoner) [inline scipy/signal/_peak_finding_utils.pyx]: Can you add a test to make sure that this is actually the behavior in flat sections?

**Review by larsoner — COMMENTED** [review]
Can you add a comment in the code about why this is better than `argrelmax` (speed and flat behavior)?

I think `float64` is fine for now, as we can expand it to other datatypes when we hit a necessary use case.

Otherwise LGTM

File: scipy/signal/tests/test_peak_finding.py
Comment (larsoner) [inline scipy/signal/tests/test_peak_finding.py]: Just noticed you have this test, ignore my comment about flat testing!

File: scipy/signal/_peak_finding_utils.pyx
Comment (perimosocordiae) [inline scipy/signal/_peak_finding_utils.pyx]: Alternately, could specify `x`'s type as a memoryview: `np.float64_t[:]`. I seem to recall people saying that was preferred for new code, but I'm not sure.

Comment (lagru) [reply scipy/signal/_peak_finding_utils.pyx]: Good point. 
As I understand it I should introduce the following changes:
```python
def _argmaxima1d(np.float64_t[:] x not None):
    ...
    cdef np.int64_t[:] maxima = np.empty(x.shape[0] // 2, dtype=np.int64)
```
Cython's doc states that [memoryviews are considerably faster](http://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#comparison-to-the-old-buffer-support). However I don't see the expected results when comparing both approaches (see [here](http://nbviewer.jupyter.org/urls/gitlab.com/snippets/1696398/raw)).

I'd be interested to know what would be the most efficient way to initialize an empty array in Cython which later needs to be accessed by Python?

I'll try profiling this later.

Comment (lagru) [reply scipy/signal/_peak_finding_utils.pyx]: Turns out this seems to be quite hard to profile as the behavior changes significantly when profiling. I think I'll leave it as introduced here 23903fb in order to avoid premature optimization. ;)

**Comment by lagru** [comment]
I think I'm happy with the current state. So how do I proceed? Should I give a short notice on the mailing list or ask a specific person to merge this?

**Review by larsoner — APPROVED** [review]
Usually a ping on GitHub is good enough. +1 for merge from me. @perimosocordiae feel free to merge if you're happy, otherwise I'll do it early next week.

**Comment by lagru** [comment]
Hey @larsoner, @perimosocordiae, I hope it's okay to post this small reminder concerning this PR.

**Comment by larsoner** [comment]
Thanks @lagru. Feel free to rebase #8264 and use this function instead.

**Comment by lagru** [comment]
> Feel free to rebase #8264 and use this function instead.

I'm a little bit confused. Wouldn't rebasing change the commits already pushed to the origin branch of #8264 which I thought is considered very bad practice? I thought I needed to merge the master with this PR into the branch of #8264...

**Comment by larsoner** [comment]
> Wouldn't rebasing change the commits already pushed to the origin branch of #8264 which I thought is considered very bad practice?

Depends on repo practice, and rationale for that practice.

If you are sharing a branch that you expect others to work on simultaneously, changing history (e.g., rebasing) can be problematic. On the other hand, rebasing maintains a cleaner history from the SciPy maintainer's / reviewer's perspective, because all of your commits when reviewing or merging come on top of some point in the history (vs being intermixed with other changes). Merge commits have at times been problematic when needing to change the history of a pull request (e.g., if a new PR needed to be started and commits were cherry-picked, etc.).

For PRs a year or so ago, maintainers often requested that PR authors squash commits down from e.g. 12 to something smaller (1 or a few generally, depending on PR size) before things could be merged. Nowadays with GitHub offering options to squash+merge it probably won't make a big difference, though.

FWIW I'd probably squash the commits down from 12 to something smaller (maybe even 1) before rebasing, too, to make life easier if there are conflicts. But again it matters a lot less now with squash+merge as a GitHub option.

**Pull request merged**
